qiskit/circuit/library/diagonal.py

```diff
@@ -0,0 +1,124 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# The structure of the code is based on Emanuel Malvetti's semester thesis at ETH in 2018,
+# which was supervised by Raban Iten and Prof. Renato Renner.
+
+
+"""Diagonal matrix circuit."""
+
+from typing import Union, List, Optional
+import cmath
+import numpy as np
+
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.circuit.exceptions import CircuitError
+
+_EPS = 1e-10
+
+
+class Diagonal(QuantumCircuit):
+    r"""Diagonal circuit.
+
+    Circuit symbol:
+
+    .. parsed-literal::
+
+             ┌───────────┐
+        q_0: ┤0          ├
+             │           │
+        q_1: ┤1 Diagonal ├
+             │           │
+        q_2: ┤2          ├
+             └───────────┘
+
+    Matrix form:
+
+    .. math::
+        DiagonalGate\ q_0, q_1, .., q_{n-1} =
+            \begin{pmatrix}
+                D[0]    & 0         & \dots     & 0 \\
+                0       & D[1]      & \dots     & 0 \\
+                \vdots  & \vdots    & \ddots    & 0 \\
+                0       & 0         & \dots     & D[n-1]
+            \end{pmatrix}
+
+    Diagonal gates are useful as representations of Boolean functions,
+    as they can map from {0,1}^2**n to {0,1}^2**n space. For example a phase
+    oracle can be seen as a diagonal gate with {+1, -1} on the diagonals. Such
+    an oracle will induce a +1 or -1 phase on the amplitude of any corresponding
+    basis state.
+
+    Diagonal gates appear in many classically hard oracular problems such as
+    Forrelation or Hidden Shift circuits.
+
+    Diagonal gates are represented and simulated more efficiently than a dense
+    2**n x 2**n unitary matrix.
+
+    The reference implementation is via the method described in
+    Theorem 7 of [1].
+
+    **Reference:**
+
+    [1] Shende et al., Synthesis of Quantum Logic Circuits, 2009
+    `arXiv:0406176 <https://arxiv.org/pdf/quant-ph/0406176.pdf>`_
+    """
+
+    def __init__(self,
+                 diag: Union[List, np.array]) -> QuantumCircuit:
+        """Create a new Diagonal circuit.
+
+        Args:
+            diag: list of the 2^k diagonal entries (for a diagonal gate on k qubits).
+
+        Returns:
+            QuantumCircuit: the diagonal gate which was attached to the circuit.
```

Comment:
I might be wrong here! But does the initializer not return `None`? Only `new` should return an instance of the class I think. Or is this added here such that it looks nice in the docs? In that case we should use this throughout Qiskit, which I'm not sure we do as of yet

Reply:
you are right